arch/x86/kvm/pkvm/vmx/ept.c

```diff
@@ -0,0 +1,203 @@
+
+	switch (pc) {
+	case _PAGE_CACHE_MODE_WB:
+		prot |= X86_MEMTYPE_WB << VMX_EPT_MT_EPTE_SHIFT;
+		break;
+	case _PAGE_CACHE_MODE_WC:
+		prot |= X86_MEMTYPE_WC << VMX_EPT_MT_EPTE_SHIFT;
+		break;
+	case _PAGE_CACHE_MODE_UC_MINUS:
+	case _PAGE_CACHE_MODE_UC:
+		prot |= X86_MEMTYPE_UC << VMX_EPT_MT_EPTE_SHIFT;
+		break;
+	case _PAGE_CACHE_MODE_WT:
+		prot |= X86_MEMTYPE_WT << VMX_EPT_MT_EPTE_SHIFT;
+		break;
+	case _PAGE_CACHE_MODE_WP:
+		prot |= X86_MEMTYPE_WP << VMX_EPT_MT_EPTE_SHIFT;
+		break;
+	default:
+		break;
+	}
+
+	return prot;
+}
+
+static int ept_pte_to_index(unsigned long vaddr, int level)
+{
+	return SPTE_INDEX(vaddr, level);
+}
+
+static unsigned long ept_level_to_size(int level)
+{
+	return KVM_HPAGE_SIZE(level);
+}
+
+static u64 ept_level_to_mask(int level)
+{
+	return (~((1UL << SPTE_LEVEL_SHIFT(level)) - 1));
+}
+
+static bool ept_pte_is_leaf(void *ptep, int level)
+{
+	return level == PG_LEVEL_4K || !ept_pte_present(ptep) || ept_pte_huge(ptep);
+}
+
+static int ept_pte_size(int level)
+{
+	return PAGE_SIZE / SPTE_ENT_PER_PAGE;
+}
+
+static int ept_pte_count(int level)
+{
+	return SPTE_ENT_PER_PAGE;
+}
+
+static void ept_set_pte(void *ptep, u64 spte)
+{
+	WRITE_ONCE(*(u64 *)ptep, spte);
+}
+
+static u64 ept_get_pte(void *ptep)
+{
+	return READ_ONCE(*(u64 *)ptep);
+}
+
+static void host_ept_flush_tlb(struct pkvm_pgtable *pgt,
+			       unsigned long vaddr, unsigned long size)
+{
+	/* TODO */
+}
+
+int pkvm_host_ept_init(struct pkvm_pgtable *pgt, void *pool_base,
+		       unsigned long pool_pages)
+{
+	unsigned long pfn = __pkvm_pa(pool_base) >> PAGE_SHIFT;
+	struct pkvm_pgtable_mm_ops mm_ops = {
+		.phys_to_virt = pkvm_phys_to_virt,
+		.virt_to_phys = pkvm_virt_to_phys,
+		.zalloc_page = host_ept_zalloc_page,
+		.get_page = host_ept_get_page,
+		.put_page = host_ept_put_page,
+		.page_count = host_ept_page_count,
+	};
+	struct pkvm_pgtable_ops pgt_ops = {
+		.pte_present = ept_pte_present,
+		.pte_huge = ept_pte_huge,
+		.pte_mkhuge = ept_pte_mkhuge,
+		.pte_to_phys = ept_pte_to_phys,
+		.pte_to_prot = ept_pte_to_prot,
+		.calc_pte_prot = ept_calc_pte_prot,
+		.vaddr_to_index = ept_pte_to_index,
+		.level_to_size = ept_level_to_size,
+		.level_to_mask = ept_level_to_mask,
+		.pte_is_leaf = ept_pte_is_leaf,
+		.pte_size = ept_pte_size,
+		.pte_count = ept_pte_count,
+		.set_pte = ept_set_pte,
+		.get_pte = ept_get_pte,
+		.flush_tlb = host_ept_flush_tlb,
+	};
+	struct pkvm_pgtable_cap cap = {
+		.level = cpu_has_vmx_ept_5levels() ? 5 : 4,
+		.allowed_pgsz = 1 << PG_LEVEL_4K,
+		.table_prot = VMX_EPT_RWX_MASK,
+	};
+	int ret;
+
+	ret = pkvm_pool_init(&host_ept_pool, pfn, pool_pages, 0);
+	if (ret)
+		return ret;
+
+	if (cpu_has_vmx_ept_2m_page())
+		cap.allowed_pgsz |=  1 << PG_LEVEL_2M;
+	if (cpu_has_vmx_ept_1g_page())
+		cap.allowed_pgsz |=  1 << PG_LEVEL_1G;
+
+	ret = pkvm_pgtable_init(pgt, cap, mm_ops, pgt_ops);
+	if (ret)
+		return ret;
+
+	pgt->root = construct_eptp(this_cpu_read(host_vcpu),
```

Comment:
Hmm... but `pgt->root` is used not by VMX hardware as EPT_POINTER but also by software, by the pKVM hypervisor, to walk the page table? What happens when pkvm_pgtable_walk() starts walking the host EPT but the physical address it uses contains some extra bits added by construct_eptp()?

Reply:
The pkvm_pgtable_walk uses pte_to_phys callback to convert the root to physical address, like I explained in the previous comment. So this should be fine.

Reply:
Yeah, it is a cheap operation. Actually, I am also fine with using root_pa, if constructing eptp when flush host ept is not a concern. Let me change this back to root_pa.